lib/ansible/inventory/manager.py

```diff
@@ -28,6 +28,7 @@
 
 from operator import attrgetter
 from random import shuffle
+from pyparsing import pyparsing_common, Word, Literal, Combine, Optional, ZeroOrMore, alphas, alphanums, nums, printables, opAssoc, infixNotation, oneOf
```

Comment:
The problem with adding new dependencies is that it limits the systems on which ansible core is installable, if we could do same w/o adding this library it is a lot more probable that we can merge this (also clog and tests would be nice).

We do have a vendoring facility but that puts an onus on maintenance on the core team side so it has to be a very compelling reason to do so.

Reply:
(I am mliu01's technical supervisor for this project.)

What about making it an optional dependency, making it available only on supported systems?
My assumption here would be that systems where you can't install pyparsing might not need as sophisticated limits as machines where you can. Not sure how bad it is to have to implementations of `--limit`.

Or otherwise what about introducing an plugin facility for `--limit`? then the current/previous limit logic would be the default plugin and users can implement their own limit plugins.

Reply:
Making it optional might work.

It does not affect just `--limit`, host patterns can be used in `hosts:` keyword and in a lookup.

I was trying to figure out how to put this into a custom inventory plugin but they are not designed to override the parsing/selection of hosts aka dealing with 'host patterns'.

Reply:
We'll try making it optional then. :)